compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1512,9 +1514,118 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         found_ty: Ty<'tcx>,
         expr: &hir::Expr<'_>,
     ) {
+        // When `expr` is `x` in something like `let x = foo.clone(); x`, need to recurse up to
+        // get `foo` and `clone`.
+        match expr.kind {
+            hir::ExprKind::Path(hir::QPath::Resolved(
+                None,
+                hir::Path { segments: [_], res: crate::Res::Local(binding), .. },
+            )) => {
+                let Some(hir::Node::Pat(hir::Pat { hir_id, .. })) = self.tcx.hir().find(*binding)
+                else {
+                    return;
+                };
+                let Some(parent) = self.tcx.hir().find(self.tcx.hir().parent_id(*hir_id)) else {
+                    return;
+                };
+
+                match parent {
+                    // foo.clone()
+                    hir::Node::Local(hir::Local { init: Some(init), .. }) => {
+                        return self.note_type_is_not_clone(diag, expected_ty, found_ty, init);
+                    }
+                    // When `expr` is more complex like a tuple
+                    hir::Node::Pat(hir::Pat { hir_id: pat_hir_id, kind: pat_kind, .. }) => {
+                        let init = match self.tcx.hir().find(self.tcx.hir().parent_id(*pat_hir_id))
+                        {
+                            Some(hir::Node::Local(hir::Local { init: Some(init), .. })) => init,
+                            _ => return,
+                        };
+
+                        match pat_kind {
+                            hir::PatKind::Tuple(pats, ..) => match init.kind {
+                                ExprKind::Tup(init_tup) => {
+                                    for (pat, init) in zip(pats.iter(), init_tup.iter()) {
+                                        if pat.hir_id == *hir_id {
+                                            return self.note_type_is_not_clone(
+                                                diag,
+                                                expected_ty,
+                                                found_ty,
+                                                init,
+                                            );
+                                        }
+                                    }
+                                }
+                                ExprKind::Block(
+                                    hir::Block {
+                                        expr:
+                                            Some(Expr {
+                                                kind: ExprKind::Tup(init_block_expr_tup), ..
+                                            }),
+                                        ..
+                                    },
+                                    _,
+                                ) => {
+                                    for (pat, init) in zip(pats.iter(), init_block_expr_tup.iter())
```

Comment:
there's a lot of rightward drift here. Perhaps you could add a method to `ExprKind` that either returns the original expression or the trailing expr of the [innermost block](https://doc.rust-lang.org/stable/nightly-rustc/rustc_hir/hir/struct.Block.html#method.innermost_block)? It could be named as something like `trailing_expr` or `innermost_expr`. if you add documentation to that it should be fine

Reply:
I agree the right drift isn't great. Not sure I see how this will help at the moment unfortunately though, maybe there's a pattern I'm not seeing, however I can't return an original expression in an `ExprKind` as it's not an `Expr`. Plus I'd need to somehow interject that into this match statement.

I'll try and have another look next week to see if I can see a way of taming the right drift.

Reply:
Yeah I think this is tricky, I think I see what you're saying but if I do something like this it's no good for the case below as I need the expr:
```
impl<'hir> ExprKind<'hir> {
    pub fn innermost_exprkind(&'hir self) -> &'hir ExprKind<'hir> {
        match self {
            ExprKind::Block(Block { expr: Some(Expr { kind, .. }), .. }, _) => kind,
            _ => self,
        }
    }
}```
On the other hand if I make it return an `Expr` to be useful then I can't just return `self` as it's the wrong type.

Reply:
Think I've managed to tame the right drift now anyway without this. Have a look at the latest when you've a second and see if it's any better. Happy to look again if further improvements can be made.